### PR TITLE
a simple example of the C++ API

### PR DESCRIPTION
**Pull request by ctb**
Ref #1364 

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make clean diff-cover` If it introduces new functionality in
  `scripts/` is it tested?
- [x] `make format diff_pylint_report cppcheck doc pydocstyle` Is it well
  formatted?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Is it documented in the `ChangeLog`?
  http://en.wikipedia.org/wiki/Changelog#Format
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?
- [x] Do the changes respect streaming IO? (Are they
  tested for streaming IO?)
- [x] Is the Copyright year up to date?

Comments welcome!

### CODE REVIEW

**Comment by standage** [comment]
Retest this please Jenkins

File: examples/c++-api/Makefile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+all: count-stuff
+
+clean:
+	-rm count-stuff 
+
+count-stuff: count-stuff.cc
+	$(CXX) -I ../../lib count-stuff.cc -o count-stuff ../../lib/liboxli.a
```
Comment (standage) [inline examples/c++-api/Makefile]: The `liboxli.a` file isn't created when I compile on OS X El Capitan. I see some `.so` files but these are for the wrong architecture.

I don't currently have access to my Ubuntu sandbox, so spinning up an Ubuntu VM for testing now.

Comment (ctb) [inline examples/c++-api/Makefile]: Try typing 'make' in the lib directory; sorry, my bad.

Comment (standage) [inline examples/c++-api/Makefile]: Ok, got it.

It doesn't look like any of the current makefile targets build `liboxli.a`. I assume this is related to your recent comments about this not currently being a priority support item.

**Comment by standage** [comment]
I get the following error when running `make` in the `example/c++-api` on Linux (Ubuntu 14.04 and Fedora 23).

```
In file included from ../../lib/hashtable.hh:59:0,
                 from ../../lib/counting.hh:49,
                 from count-stuff.cc:7:
../../lib/traversal.hh:75:37: error: ‘std::function’ has not been declared
                                std::function<bool (Kmer&)> filter);
                                     ^
../../lib/traversal.hh:75:45: error: expected ‘,’ or ‘...’ before ‘<’ token
                                std::function<bool (Kmer&)> filter);
                                             ^
../../lib/traversal.hh:78:38: error: ‘std::function’ has not been declared
                                 std::function<bool (Kmer&)> filter);
                                      ^
../../lib/traversal.hh:78:46: error: expected ‘,’ or ‘...’ before ‘<’ token
                                 std::function<bool (Kmer&)> filter);
                                              ^
make: *** [count-stuff] Error 1
```

**Comment by luizirber** [comment]
Need to set cpp standard to c++11

**Comment by standage** [comment]
Ok thanks. Adding `--std=c++11` did the trick.

**Comment by ctb** [comment]
(Please do go ahead and fix this branch!)

**Comment by standage** [comment]
I assume I'm free to merge this once the tests pass again. Or am I disqualified now that I've pushed a commit?

**Comment by ctb** [comment]
Best wait... Is there more we want to do here? Link to docs? Etc

--
Titus Brown, ctbrown@ucdavis.edu

**Comment by standage** [comment]
In that vein...this is our first developer-facing example, and we have only a single user-facing example. It's probably not worth separating these into different directories in the code distribution, but we should probably consider this for the docs.

**Comment by ctb** [comment]
As you say... Also, checking that these work should be added into release checklist. 

For shell examples, we should link in recipes and protocols (in other repos) and maybe for Python examples we could make a binder thingy on mybinder.

Infinite work.... :)

--
Titus Brown, ctbrown@ucdavis.edu

File: examples/c++-api/count-stuff.cc [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    tablesize.push_back(pow(4, ksize));
+    
+    CountingHash ktable(ksize, tablesize);
+
+    ktable.consume_string("ATGGCGATGGCAAGTAGGACCCAGATGGACCAAAG");
+
+    std::cout << "count for: " << "ATGGCGATGGC" << " is " <<
+        ktable.get_count("ATGGCGATGGC") << "\n";
+
+    ktable.count("ATGGCGATGGC");
```
Comment (standage) [inline examples/c++-api/count-stuff.cc]: Perhaps a comment here about the behavior of `get_count()` with vs without a preceding `count()` call is warranted?

Comment (standage) [inline examples/c++-api/count-stuff.cc]: Uh, I erroneously assumed `count` was similar in behavior to `get_count`. Seeing #1380 cleared this up for me, that `count` is an alias for `add`.

Comment (ctb) [inline examples/c++-api/count-stuff.cc]: Exactly why I wanted to change it - made that mistake myself :). Can you update example?

Comment (standage) [inline examples/c++-api/count-stuff.cc]: Sure thang.

**Comment by standage** [comment]
It looks like the `add` function was only implemented at the Python level. I had to add it at the C++ level for this example.

**Comment by standage** [comment]
Added to docs, and updated release checklist. Could probably use some filling out, but that would be a separate issue.

Ready for review @ctb.

**Comment by standage** [comment]
retest this please jenkins

File: examples/c++-api/count-stuff.cc [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#include <vector>
+#include <cmath>
+#include "khmer.hh"
+#include "counting.hh"
+
+using namespace khmer;
+
+int main()
+{
+    unsigned int ksize = 11;
```
Comment (ctb) [inline examples/c++-api/count-stuff.cc]: Might want to mention that ksize should be odd; if it's even, then reverse complements will collapse into one entry, and you'll want to use a tablesize of ... what, 4**(k-1) + k, or something?

**Comment by standage** [comment]
Retest this please Jenkins

**Comment by standage** [comment]
Any more comments @ctb?

**Comment by ctb** [comment]
LGTM - think it's ready to merge?

**Comment by standage** [comment]
@ctb Yessir

**Pull request merged**